rcl/src/rcl/time.c

```diff
@@ -12,8 +12,308 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <stdlib.h>
+
 #if defined(WIN32)
 #include "./time_win32.c"
 #else  // defined(WIN32)
 #include "./time_unix.c"
 #endif  // defined(WIN32)
+
+#include "./stdatomic_helper.h"
+
+// Process default ROS time sources
+static rcl_time_source_t * rcl_default_ros_time_source;
+static rcl_time_source_t * rcl_default_steady_time_source;
+static rcl_time_source_t * rcl_default_system_time_source;
+
+
+/** Internal storage for RCL_ROS_TIME implementation */
+typedef struct
+{
+  atomic_uint_least64_t current_time;
+  bool active;
+  // TODO(tfoote): store subscription here
+} rcl_ros_time_source_storage_t;
+
+/** Implementation only **/
+rcl_ret_t rcl_get_steady_time(void * data, rcl_time_point_imp_t * current_time)
+{
+  (void) data; /* unused */
+  return rcl_steady_time_now(current_time);
+}
+
+/** Implementation only **/
+rcl_ret_t rcl_get_system_time(void * data, rcl_time_point_imp_t * current_time)
+{
+  (void) data; /* unused */
+  return rcl_system_time_now(current_time);
+}
+
+/* Internal method for zeroing values on init
+ * Assumes time_source is valid*/
+void rcl_init_generic_time_source(rcl_time_source_t * time_source)
+{
+  time_source->type = RCL_TIME_SOURCE_UNINITIALIZED;
+  time_source->pre_update = NULL;
+  time_source->post_update = NULL;
+  time_source->get_now = NULL;
+  time_source->data = NULL;
+}
+
+/* The function used to get the current ros time.
+ * This is in the implementation only.*/
+rcl_ret_t rcl_get_ros_time(void * data, rcl_time_point_imp_t * current_time)
+{
+  rcl_ros_time_source_storage_t * t = (rcl_ros_time_source_storage_t *)data;
+  if (!t->active) {
+    return rcl_get_system_time(data, current_time);
+  }
+  *current_time = rcl_atomic_load_uint64_t(&(t->current_time));
+  return RCL_RET_OK;
+}
+
+bool rcl_time_source_valid(rcl_time_source_t * time_source)
+{
+  if (time_source == NULL ||
+    time_source->type == RCL_TIME_SOURCE_UNINITIALIZED ||
+    time_source->get_now == NULL)
+  {
+    return false;
+  }
+  return true;
+}
+
+rcl_ret_t rcl_init_ros_time_source(rcl_time_source_t * time_source)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(time_source, RCL_RET_INVALID_ARGUMENT);
+  rcl_init_generic_time_source(time_source);
+  time_source->data = calloc(1, sizeof(rcl_ros_time_source_storage_t));
+  time_source->get_now = rcl_get_ros_time;
+  time_source->type = RCL_ROS_TIME;
+  if (!time_source) {
```

Comment:
How can this be 0 if it has been used already in the three lines before? Does the check need to be moved higher? Also it doesn't seem to set any error message?

Same below.